diff --git a/services/core/java/com/android/server/wm/WindowManagerService.java b/services/core/java/com/android/server/wm/WindowManagerService.java
index ffdf18a..b5fce94 100644
--- a/services/core/java/com/android/server/wm/WindowManagerService.java
+++ b/services/core/java/com/android/server/wm/WindowManagerService.java
@@ -4841,7 +4841,12 @@ public class WindowManagerService extends IWindowManager.Stub
                 if (mAppsFreezingScreen == 1) {
                     startFreezingDisplayLocked(false, 0, 0);
                     mH.removeMessages(H.APP_FREEZE_TIMEOUT);
-                    mH.sendEmptyMessageDelayed(H.APP_FREEZE_TIMEOUT, 2000);
+                    int timeoutMs = 2000;
+                    if (wtoken.toString().contains("com.android.contacts/.activities.PeopleActivity")) {
+                        Slog.i(TAG, "zzj   app free time    test   peopleActivity");
+                        timeoutMs = 5000;
+                    }
+                    mH.sendEmptyMessageDelayed(H.APP_FREEZE_TIMEOUT, timeoutMs);
                 }
             }
             final int N = wtoken.allAppWindows.size();
