diff --git a/core/java/android/view/ViewRootImpl.java b/core/java/android/view/ViewRootImpl.java
index 6e93a30..64c6ff4 100644
--- a/core/java/android/view/ViewRootImpl.java
+++ b/core/java/android/view/ViewRootImpl.java
@@ -216,6 +216,7 @@ public final class ViewRootImpl implements ViewParent,
     boolean mReportNextDraw;
     boolean mFullRedrawNeeded;
     boolean mNewSurfaceNeeded;
+    boolean mTmpRelayoutNeeded;
     boolean mHasHadWindowFocus;
     boolean mLastWasImTarget;
     boolean mWindowsAnimating;
@@ -828,6 +829,7 @@ public final class ViewRootImpl implements ViewParent,
 
     void handleAppVisibility(boolean visible) {
         if (mAppVisible != visible) {
+        	Log.e(TAG, "zzj   XXXXXXXXXXX    handleAppVisibility="  + visible + "  " + mWindowAttributes.getTitle());
             mAppVisible = visible;
             scheduleTraversals();
             if (!mAppVisible) {
@@ -1536,7 +1538,12 @@ public final class ViewRootImpl implements ViewParent,
         int relayoutResult = 0;
 
         if (mFirst || windowShouldResize || insetsChanged ||
-                viewVisibilityChanged || params != null) {
+                viewVisibilityChanged || params != null ||
+                (mTmpRelayoutNeeded && viewVisibility == View.VISIBLE)) {
+            if (!(mFirst || windowShouldResize || insetsChanged ||
+                viewVisibilityChanged || params != null) && (mTmpRelayoutNeeded && viewVisibility == View.VISIBLE)) {
+				Log.i(TAG, "zzjxx        XXXXXXXX     give another change to relayoutWindow " + mWindowAttributes.getTitle());
+            }
 
             if (viewVisibility == View.VISIBLE) {
                 // If this window is giving internal insets to the window
@@ -1752,6 +1759,10 @@ public final class ViewRootImpl implements ViewParent,
                             mAttachInfo.mHardwareRenderer.isEnabled()) {
                         mAttachInfo.mHardwareRenderer.destroy();
                     }
+                    if(viewVisibility == View.VISIBLE) {
+                    	Slog.e("zzj", "      XXXXXXXXXXx     mTmpRelayoutNeeded = true;    " + mWindowAttributes.getTitle());
+						mTmpRelayoutNeeded = true;
+                    }
                 } else if (surfaceGenerationId != mSurface.getGenerationId() &&
                         mSurfaceHolder == null && mAttachInfo.mHardwareRenderer != null) {
                     mFullRedrawNeeded = true;
@@ -2031,6 +2042,7 @@ public final class ViewRootImpl implements ViewParent,
         mWillDrawSoon = false;
         mNewSurfaceNeeded = false;
         mViewVisibility = viewVisibility;
+        mTmpRelayoutNeeded = false;
 
         if (mAttachInfo.mHasWindowFocus && !isInLocalFocusMode()) {
             final boolean imTarget = WindowManager.LayoutParams
@@ -2070,6 +2082,11 @@ public final class ViewRootImpl implements ViewParent,
             if (viewVisibility == View.VISIBLE) {
                 // Try again
                 scheduleTraversals();
+                //the surface may not created beacuse of server's win.mAppToken.clientHidden set true
+                //We set this flag to give another chance to create surface
+                //if (!newSurface) {
+                //    mTmpRelayoutNeeded = true;
+                //}
             } else if (mPendingTransitions != null && mPendingTransitions.size() > 0) {
                 for (int i = 0; i < mPendingTransitions.size(); ++i) {
                     mPendingTransitions.get(i).endChangingAnimations();
